Properly verify PrepareRequest fields

AnnaShaleva opened the pull request:
Close https://github.com/bane-labs/go-ethereum/issues/53. Related to #65 and #87.

This PR introduces PrepareRequest verification on `WithVerifyPrepareRequest` callback level. It reuses as much code of `verifyHeader` as possible.

AnnaShaleva committed:
dbft: properly verify PrepareRequest fields

Close #53. Verify those block fields that won't be filled in from dBFT
context.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: add direct Extra version check to verifyHeader

This check is present in `getSignersAndSigs`, but we'd better ensure
that Extra has proper (known) version at the start of block verification,
before witness checks.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@nicolegys, @chenquanyu, you are welcomed to review&test.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -688,14 +689,14 @@ func (c *DBFT) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types
-	// Don't waste time checking blocks from the future
-	if header.Time > uint64(time.Now().Unix()) {
+	// Don't waste time checking blocks from the future.
+	if isSealed && header.Time > uint64(time.Now().Unix()) {
```
We can't accept _any_ time in proposal.

AnnaShaleva replied:
We do check the proposal time, the condition is not to be lower than the previous block. This check is located at the `WithVerifyBlock` level, see https://github.com/bane-labs/go-ethereum/pull/87#discussion_r1472866452. I agree that this check can be done at the PrepareRequest verification level, but currently it's done exactly like in NeoGo, and it works properly (so *any* timestamp can't be accepted by CNs, only larger than the previous block's timestamp).

roman-khimov replied:
OK, that will do the trick. The most important thing is for it to be checked _before_ sending PrepareResponse.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -710,29 +711,32 @@ func (c *DBFT) verifyHeader(chain consensus.ChainHeaderReader, header *types.Hea
+		}
+
+		// Ensure that the mix digest is not zero.
+		if header.MixDigest == (common.Hash{}) {
```
Can we have it in proposal and check local one against proposed?

AnnaShaleva replied:
We can, but why do we need this? MixDigest depends on validators list, nodes take validators list from the dBFT context, and if something is wrong with the validators list, the consensus just won't work. And nodes fill this field by their own before commit is send to include this field in the hashable part, exactly like in NeoGo. If malicious node wants to change this field, then Commit verification will fail on a good node. If malicios node will do everything right and intentionally change MixDigest after Commit release, then the resulting block relayed to the network by this node won't be accepted by other nodes because it has wrong MixDigest. But blocks relayed by other honest consensus nodes will be accepted.

So I don't think we need to include this field in PrepareRequest. And its stub will be removed from PrepareRequest in #82. Am I missing something here?

roman-khimov replied:
Fail fast. And better with some meaningful reason. Right that would be "wrong commit signature", go figure.

AnnaShaleva replied:
Should we perform the same update in NeoGo?

AnnaShaleva replied:
> Should we perform the same update in NeoGo?

OK, silly question, can't do this in NeoGo since NextConsensus is not a part of PrepareRequest and protocol extension is needed there. OK, will do it in Geth at least.

AnnaShaleva replied:
Fixed in a separate commit. But we need to recheck it once Governance contract is integrated and state-dependent validators calculations are available.

AnnaShaleva replied:
See https://github.com/bane-labs/go-ethereum/issues/14#issuecomment-1923566099.

roman-khimov replied:
It's going to be tested wrt validator list change anyway, I think.

AnnaShaleva committed:
dbft: include NextConsensus to PrepareRequest

Include NextConsensus field to PrepareRequest andverify it during PrepareRequest
verification. Compute next block NextConsensus address once after current block
persist and store it in cache.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, ready for review.
